src/mds/Server.cc

```diff
@@ -465,6 +465,11 @@ void Server::handle_client_session(MClientSession *m)
       mdlog->flush();
     break;
 
+  case CEPH_SESSION_RECALL_STATEMSG_ACK:
+    if (this->report_acks)
+      session->recall_acked = true;
```

Comment:
The right way to do this is to record a sequence number for the recall messages the MDS sends so that the client can ACK the seq number. Then we know which recall messages it has handled. Recall what Zheng said in the email thread: "we can add a recall_state seq to MClientCapRelease and MClientSession message.".

Please modify those messages (and increment the message version) to include the sequence number.

Reply:
Please correct me if I am wrong: in my understanding, I need to assign seq. numbers to `CEPH_SESSION_RECALL_STATE` messages being sent by MDS in `Server::client_recall_state`, store them for later and then while receiving ACKs (until `mds_session_recall_ack_timeout`) for these messages I need to check them and report the ones (in the logs) I haven't received an ACK for.

> Recall what Zheng said in the email thread: "we can add a recall_state seq to MClientCapRelease and MClientSession message.".

Yeah, I couldn't comprehend the "seq" part back then.

Reply:
Updating the PR as per my understanding above.